### PR TITLE
Mark slowest tests as slow and add nightly slow test job

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit makes two changes to hopefully improve CI throughput. The first is to mark the majority of the shor's algorithm testing and some other particularly slowly executing as slow. This prevents the tests from running as part of normal CI runs which should hopefully improve the runtime of a full unit test job. To make sure we don't forget about these tests and regress the functionality this commit also adds a new nightly CI job to github actions to run all the tests including the ones decorated with `@slow_test`. If the nightly job fails we can track it with an issue and fix it.

### Details and comments

### CODE REVIEW

**Review by jakelishman — COMMENTED** [review]
I support doing this - in particular, removing the (deprecated) Shor and H2 tests from the PR CI pipelines should have a really meaningful impact on the critical path.  It's not clear exactly how much the parallelisation gets invoked over the most expensive tests, but we're cutting out ~10 minutes of tests and our Linux VMs get 2x parallelisation, so this could be up to a 20-minute reduction per commit in a merge train (plus whatever effects we get from reduction in queue congestion from other CI finishing sooner).

File: .github/workflows/slow.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        name: Install Python
+        with:
+          python-version: '3.10'
+      - name: Install dependencies
+        run: |
+          pip install -U -r requirements.txt -c constraints.txt
+          pip install -U -r requirements-dev.txt coveralls -c constraints.txt
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply .github/workflows/slow.yml:19]: It's just sloppy copy and paste I'll remove this line

**Review by ajavadia — APPROVED** [review]

**Comment by jakelishman** [comment]
@ajavadia: The PR just needs a couple of minor updates above to cut out some unnecessary bits / naming confusion in the workflow file, but otherwise we're ready to merge.

File: .github/workflows/slow.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtreinish) [reply .github/workflows/slow.yml]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/8811/commits/28d3d14cfcc25b9e5fe8e58acd1aefee52d0d2ed

File: .github/workflows/slow.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+name: Slow nightly tests
+on:
+  schedule:
+    - cron:  '42 3 * * *'
+  workflow_dispatch:
+jobs:
+  randomized_tests:
+    name: Full-test-run-with-slow
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply .github/workflows/slow.yml]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/8811/commits/28d3d14cfcc25b9e5fe8e58acd1aefee52d0d2ed

**Review by jakelishman — APPROVED** [review]

**Pull request merged**